packages/flutter/lib/src/material/toggle_buttons.dart

```diff
@@ -0,0 +1,951 @@
+          onPressed: onPressed,
+          child: child,
+        ),
+      ),
+    );
+
+    return _SelectToggleButton(
+      key: key,
+      leadingBorderSide: leadingBorderSide,
+      horizontalBorderSide: horizontalBorderSide,
+      trailingBorderSide: trailingBorderSide,
+      borderRadius: borderRadius,
+      isFirstButton: isFirstButton,
+      isLastButton: isLastButton,
+      child: result,
+    );
+  }
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(FlagProperty('selected',
+      value: selected,
+      ifTrue: 'Button is selected',
+      ifFalse: 'Button is unselected',
+    ));
+  }
+}
+
+class _SelectToggleButton extends SingleChildRenderObjectWidget {
+  const _SelectToggleButton({
+    Key key,
+    Widget child,
+    this.leadingBorderSide,
+    this.horizontalBorderSide,
+    this.trailingBorderSide,
+    this.borderRadius,
+    this.isFirstButton,
+    this.isLastButton,
+  }) : super(
+    key: key,
+    child: child,
+  );
+
+  // The width and color of the button's leading side border.
+  final BorderSide leadingBorderSide;
+
+  // The width and color of the button's top and bottom side borders.
+  final BorderSide horizontalBorderSide;
+
+  // The width and color of the button's trailing side border.
+  final BorderSide trailingBorderSide;
+
+  // The border radii of each corner of the button.
+  final BorderRadius borderRadius;
+
+  // Whether or not this toggle button is the first button in the list.
+  final bool isFirstButton;
+
+  // Whether or not this toggle button is the last button in the list.
+  final bool isLastButton;
+
+  @override
+  _SelectToggleButtonRenderObject createRenderObject(BuildContext context) => _SelectToggleButtonRenderObject(
+    leadingBorderSide: leadingBorderSide,
+    horizontalBorderSide: horizontalBorderSide,
+    trailingBorderSide: trailingBorderSide,
+    borderRadius: borderRadius,
+    isFirstButton: isFirstButton,
+    isLastButton: isLastButton,
+    textDirection: Directionality.of(context),
+  );
+
+  @override
+  void updateRenderObject(BuildContext context, _SelectToggleButtonRenderObject renderObject) {
+    renderObject
+      ..leadingBorderSide = leadingBorderSide
+      ..horizontalBorderSide = horizontalBorderSide
+      ..trailingBorderSide = trailingBorderSide
+      ..borderRadius = borderRadius
+      ..isFirstButton = isFirstButton
+      ..isLastButton = isLastButton
+      ..textDirection = Directionality.of(context);
+  }
+}
+
+class _SelectToggleButtonRenderObject extends RenderShiftedBox {
+  _SelectToggleButtonRenderObject({
+    this.leadingBorderSide,
+    this.horizontalBorderSide,
+    this.trailingBorderSide,
+    this.borderRadius,
+    this.isFirstButton,
+    this.isLastButton,
+    this.textDirection,
+    RenderBox child,
+  }) : super(child);
+
+  // The width and color of the button's leading side border.
+  BorderSide leadingBorderSide;
```

Comment:
Each of the renderer's properties should be defined like this.

```
int get foo => _foo;
int _foo;
set foo(int value) {
  if (value == _foo) {
    return;
  }
  _foo = value;
  markNeedsPaint(); // or markNeedsLayout()
}
```

Verify that the these methods are doing the right thing by writing tests that rebuild ToggleButtons with new foo values; check what's painted and check that the layout is still right.